REGRESSION (UI-side compositing): Scrolling web inspector timeline is very jumpy

nmoucht opened the pull request:
#### 1bb52fdceac9130525dcb521bb779ee184a392eb
<pre>
REGRESSION (UI-side compositing): Scrolling web inspector timeline is very jumpy
<a href="https://bugs.webkit.org/show_bug.cgi?id=258319">https://bugs.webkit.org/show_bug.cgi?id=258319</a>
rdar://107685436

Reviewed by Simon Fraser.

When receiving a wheel event from dispatchEvent, the current design recieves this event
in the web process, and sends this request to the UI-process as a RequestedScrollData
object. This object contained a scrollPosition to animate to, but this was calculated
using the scroll position in the web process. Since this scroll position could be out
of sync with the UI-process while the user was scrolling, this could result in a scroll
to a scroll position behind the current scroll position, leading to a jumpy looking scroll.
To fix this, rather than sending across the scroll position to scroll to, simply send across
the requested scroll delta to the UI-process, and use the UI-process scroll position to
calculate the correct scroll position to animate to. Also, propagate down the original
scroll delta (if it exists), to be used in AsyncScrollingCoordinator::requestScrollToPosition,
so that we know whether to do a delta or scroll position update.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::startPendingScrollAnimations):

Canonical link: <a href="https://commits.webkit.org/266210@main">https://commits.webkit.org/266210@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cee341438d14c78d549824ad866283b45708d494

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13169 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13482 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14907 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12545 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15990 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13509 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15238 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14015 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15440 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11303 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11891 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18950 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12378 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12058 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15271 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12541 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10415 "2 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11811 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16132 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1502 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1164,15 +1164,15 @@ void Element::scrollBy(const ScrollToOptions& options)
     scrollToOptions.left.value() += scrollLeft();
     scrollToOptions.top.value() += scrollTop();
-    scrollTo(scrollToOptions, ScrollClamping::Clamped, ScrollSnapPointSelectionMethod::Directional);
+    scrollTo(scrollToOptions, ScrollClamping::Clamped, ScrollSnapPointSelectionMethod::Directional, ScrollClass::ScrollBy);
```
As far as I can tell, `scrollToOptions` at this point contains a specific destination, so this is is a `ScrollTo` at this point. Is there any case in which `scrollToOptions.left` represents a delta, not an absolute value?

nmoucht replied:
This was intended to be passed down into AsyncScrollingCoordinator::requestScrollToPosition to indicate whether we should send across the delta or scroll position. This has been made more clear by passing down the original delta in the options as well.

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -197,7 +197,7 @@ class Element : public ContainerNode {
     void scrollBy(const ScrollToOptions&);
     void scrollBy(double x, double y);
-    virtual void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest);
+    virtual void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest, ScrollClass scrollClass = ScrollClass::ScrollTo);
```
This is explicitly a `scrollTo` so I'm not sure how the parameter makes sense.

nmoucht replied:
See above.

smfr commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1770,15 +1770,15 @@ void LocalDOMWindow::scrollBy(const ScrollToOptions& options) const
     scrollToOptions.left.value() += view->mapFromLayoutToCSSUnits(view->contentsScrollPosition().x());
     scrollToOptions.top.value() += view->mapFromLayoutToCSSUnits(view->contentsScrollPosition().y());
-    scrollTo(scrollToOptions, ScrollClamping::Clamped, ScrollSnapPointSelectionMethod::Directional);
+    scrollTo(scrollToOptions, ScrollClamping::Clamped, ScrollSnapPointSelectionMethod::Directional, ScrollClass::ScrollBy);
```
Same comment as before.

nmoucht replied:
See above.

smfr commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -277,7 +277,7 @@ class LocalDOMWindow final
     void scrollBy(const ScrollToOptions&) const;
     void scrollBy(double x, double y) const;
-    void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest) const;
+    void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest, ScrollClass scrollClass = ScrollClass::ScrollTo) const;
```
This is explicitly a `scrollTo` so I'm not sure how the parameter makes sense.

nmoucht replied:
See above.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -341,7 +339,13 @@ bool AsyncScrollingCoordinator::requestScrollToPosition(ScrollableArea& scrollab
 
-    stateNode->setRequestedScrollData({ ScrollRequestType::PositionUpdate, scrollPosition, scrollType, clamping, animated });
+    if (options.animated == ScrollIsAnimated::Yes && options.scrollClass == ScrollClass::ScrollBy)
+        stateNode->setRequestedScrollData({ ScrollRequestType::DeltaUpdate, { }, scrollPosition - scrollableArea.scrollPosition(), options.type, options.clamping, options.animated });
```
Not really a fan of computing the delta like this; it would be cleaner to just pass the delta down, but maybe that touches too many functions?

nmoucht replied:
Yeah this was the idea behind using the scroll class to indicate if the scroll to position was originally a delta or a position to scroll to. Rather than subtracting the current scroll position for the delta case, I think passing the original scroll delta down through `ScrollPositionChangeOptions` makes more sense.

nmoucht replied:
Added originalScrollDelta to `ScrollPositionChangeOptions`.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
```diff
@@ -47,6 +47,11 @@ void RequestedScrollData::merge(RequestedScrollData&& other)
     *this = WTFMove(other);
 }
 
+FloatPoint RequestedScrollData::scrollToPosition(const FloatPoint& currentScrollPosition) const
```
Confusing name. Maybe `destinationPosition`?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
```diff
@@ -116,18 +116,21 @@ enum class ViewportRectStability {
 struct RequestedScrollData {
     ScrollRequestType requestType { ScrollRequestType::PositionUpdate };
     FloatPoint scrollPosition;
+    FloatSize scrollDelta { };
```
No need for {}. Since `scrollPosition` and `scrollDelta` should never both be set, perhaps this should use a `std::variant<>`.

nmoucht replied:
Fixed to just be scrollPosition.

smfr commented on Source/WebCore/platform/ScrollTypes.h:
```diff
@@ -341,35 +341,30 @@ enum class ScrollSnapPointSelectionMethod : uint8_t {
     Closest,
 };
 
+enum class ScrollClass : uint8_t {
```
Maybe `ScrollType` ?

nmoucht replied:
Unfortunately `ScrollType` is being used for differentiating user vs programatic scroll.

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -197,7 +197,7 @@ class Element : public ContainerNode {
     void scrollBy(const ScrollToOptions&);
     void scrollBy(double x, double y);
-    virtual void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest);
+    virtual void scrollTo(const ScrollToOptions&, ScrollClamping = ScrollClamping::Clamped, ScrollSnapPointSelectionMethod = ScrollSnapPointSelectionMethod::Closest, ScrollClass scrollClass = ScrollClass::ScrollTo, FloatPoint originalScrollDelta = { });
```
Scroll delta should be FloatSize.

smfr replied:
Also originalScrollDelta is only non-zero when ScrollClass is ScrollBy, so there's redundancy here. You could use std::optional<FloatSize> to indicate it's a delta, and remove ScrollClass?

nmoucht replied:
Fixed to use a `std::optional<FloatSize>`.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -341,7 +341,13 @@ bool AsyncScrollingCoordinator::requestScrollToPosition(ScrollableArea& scrollab
         return false;
 
-    stateNode->setRequestedScrollData({ ScrollRequestType::PositionUpdate, scrollPosition, scrollType, clamping, animated });
+    if (options.animated == ScrollIsAnimated::Yes && options.scrollClass == ScrollClass::ScrollBy)
```
Maybe we should always do the scrollBy thing for ScrollClass::ScrollBy, and not just when animated.

nmoucht replied:
Right I agree.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
```diff
@@ -34,7 +34,7 @@ void RequestedScrollData::merge(RequestedScrollData&& other)
         // below in the case where the cancelled animated scroll is immediately followed by another animated scroll.
         other.scrollPosition = scrollPosition;
-    } else if (other.requestType == ScrollRequestType::PositionUpdate && other.animated == ScrollIsAnimated::Yes) {
+    } else if ((other.requestType == ScrollRequestType::PositionUpdate || other.requestType == ScrollRequestType::DeltaUpdate) && other.animated == ScrollIsAnimated::Yes) {
```
I'm not sure this merging is correct. If there's a requested ScrollRequestType::PositionUpdate already pending, and a DeltaUpdate comes in, should you add the delta to the position request? Can you make a test that exercises this code?

nmoucht replied:
I don't think it's correct to add a DeltaUpdate to a PositionUpdate. These two should be functionally equivalent to each other, so they should use the same existing merge behavior, where a new animated scroll overwrites the previous animated scroll (and passes forward the `requestedDataBeforeAnimatedScroll` if it exists), and a non animated scroll simply overwrites the current RequestedScrollData. This merge code is pretty well covered by current tests (like those in imported/w3c/web-platform-tests/css/cssom-view/scroll-behavior-*).

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
```diff
@@ -47,6 +47,11 @@ void RequestedScrollData::merge(RequestedScrollData&& other)
 
+FloatPoint RequestedScrollData::destinationPosition(const FloatPoint& currentScrollPosition) const
+{
+    return requestType == ScrollRequestType::DeltaUpdate ? (currentScrollPosition + scrollPosition) : scrollPosition;
```
Not a fan of how `position` for a delta upate is really a delta.

nmoucht replied:
Fixed to use a `std::variant<FloatSize, FloatPoint>`.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
```diff
@@ -116,23 +116,33 @@ enum class ViewportRectStability {
 
     void merge(RequestedScrollData&&);
+    WEBCORE_EXPORT FloatPoint destinationPosition(const FloatPoint&) const;
+    bool comparePositionOrDelta(const RequestedScrollData& other) const
```
"compare" doesn't tell the reader if a true return value means "matched" or "mismatched", so it would be better to name it `positionOrDeltaEqual()`

nmoucht committed:
REGRESSION (UI-side compositing): Scrolling web inspector timeline is very jumpy
https://bugs.webkit.org/show_bug.cgi?id=258319
rdar://107685436

Reviewed by Simon Fraser.

When receiving a wheel event from dispatchEvent, the current design recieves this event
in the web process, and sends this request to the UI-process as a RequestedScrollData
object. This object contained a scrollPosition to animate to, but this was calculated
using the scroll position in the web process. Since this scroll position could be out
of sync with the UI-process while the user was scrolling, this could result in a scroll
to a scroll position behind the current scroll position, leading to a jumpy looking scroll.
To fix this, rather than sending across the scroll position to scroll to, simply send across
the requested scroll delta to the UI-process, and use the UI-process scroll position to
calculate the correct scroll position to animate to. Also, propagate down the original
scroll delta (if it exists), to be used in AsyncScrollingCoordinator::requestScrollToPosition,
so that we know whether to do a delta or scroll position update.

* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::startPendingScrollAnimations):

Canonical link: https://commits.webkit.org/266210@main
